src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -26,25 +43,107 @@
 
 @WebServlet("/search")
 public class SearchServlet extends HttpServlet {
-  
-  /*
-  * doGet receives the request and returns the message sent as parameter and a hardcoded list of Recipes
-  */
+
+  /**
+   * Search and returns a list of first 10 recipes with the title matching the given parameter(searchterm).
+   * The index returns a list of documents in the ascending order by title
+   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
     throws ServletException, IOException {
-    request.setAttribute("messages", request.getParameter("searchterm"));
-
-    ArrayList<Recipe> recipes = new ArrayList<>();
-    for (int i = 0; i < 5; ++i) {
-      Recipe recipe = new Recipe();
-      recipe.setId(i);
-      recipe.setName("Cake");
-      recipe.setDescription("Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.");
-      recipes.add(recipe);
+    List<Recipe> recipesList = new ArrayList<>();
+    String titleToMatch = request.getParameter("searchterm");
+    Index searchIndex = getIndex();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query query = buildQuery(titleToMatch);
+    Results<ScoredDocument> resultsByTitle = searchIndex.search(query);
+
+    for (ScoredDocument entity : resultsByTitle) {
+      try {
+        Entity recipeEntity = getRecipeEntityFromDocumentEntity(
+          datastore,
+          entity
+        );
+        Recipe recipe = buildRecipe(recipeEntity);
+
+        recipesList.add(recipe);
+      } catch (EntityNotFoundException e) {
+        response.setStatus(505);
```

Comment:
You should do

 request.getRequestDispatcher("/search.jsp").forward(request, response);
return;

Here, you don't want to return any data if there was an error.